### PR TITLE
[perms] Grant owner/member on role change WEB-505, WEB-507

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Adds owner/member relations to users, when they either:
* Create Org
* Change Role of a user

## Related Issue(s)

* [WEB-505](https://linear.app/gitpod/issue/WEB-505/migrate-createteam)
* [WEB-507](https://linear.app/gitpod/issue/WEB-507/migrate-setteammemberrole)

## How to test

* Preview
* Change role of a user
* Deletion of users does not yet remove the relation

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - mp-perms-cc78c52d95d</li>
	<li><b>🔗 URL</b> - <a href="https://mp-perms-cc78c52d95d.preview.gitpod-dev.com/workspaces" target="_blank">mp-perms-cc78c52d95d.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/server/src/authorization/relationships.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    OrganizationPermission,
+    OrganizationResourceType,
+    UserResourceType,
+    objectRef,
+    relationship,
+    subject,
+} from "./definitions";
+import { v1 } from "@authzed/authzed-node";
+
+export function organizationOwnerRole(requestor: string, orgID: string, userID: string): v1.WriteRelationshipsRequest {
```
Comment (geropl) [inline components/server/src/authorization/relationships.ts:17]: :heart: The per-usecase method :upside_down_face:

**Comment by easyCZ** [comment]
Looking for ideas on how to add unit tests to this, given our constraints. See [internal thread](https://gitpod.slack.com/archives/C02EN94AEPL/p1687356620905269)

File: components/gitpod-db/src/typeorm/team-db-impl.ts [hunk 2]
```diff
@@ -186,20 +186,15 @@ export class TeamDBImpl extends TransactionalDBImpl<TeamDB> implements TeamDB {
             throw new ResponseError(ErrorCodes.BAD_REQUEST, "Please choose a name that is at most 64 characters long.");
         }
 
-        // Storing new entry in a TX to avoid potential dupes caused by racing requests.
-        const em = await this.getEntityManager();
```
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/team-db-impl.ts:190]: Need to remove the migration from here, as we are already wrapping it with one. The `gitpod-service` is the only call-site that I could find.

File: components/server/src/authorization/relationships.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+export function removeUserFromOrg(orgID: string, userID: string): v1.WriteRelationshipsRequest {
+    return v1.WriteRelationshipsRequest.create({
+        updates: [
+            v1.RelationshipUpdate.create({
+                operation: v1.RelationshipUpdate_Operation.DELETE,
+                relationship: relationship(objectRef("organization", orgID), "member", subject("user", userID)),
+            }),
+            v1.RelationshipUpdate.create({
+                operation: v1.RelationshipUpdate_Operation.DELETE,
+                relationship: relationship(objectRef("organization", orgID), "member", subject("user", userID)),
```
Comment (geropl) [inline components/server/src/authorization/relationships.ts:61]:
```suggestion
                relationship: relationship(objectRef("organization", orgID), "owner", subject("user", userID)),
```
Nice try, @easyCZ! :wink:

Comment (easyCZ) [reply components/server/src/authorization/relationships.ts]: Fixed. Thanks. We'll be circling back with unit tests once we add an OrganizationService (Alex is looking at this) which allows us to test this for RPCs

**Comment by geropl** [comment]
Tested: When trying to sign-up, and clicking the "Continue" button on the questionaire, I get `permission_denied] No access to Organization ID: 2a511bc8-cf3d-4911-a540-6169c5eea39f` on the console.

**Comment by easyCZ** [comment]
> Tested: When trying to sign-up, and clicking the "Continue" button on the questionaire, I get `permission_denied] No access to Organization ID: 2a511bc8-cf3d-4911-a540-6169c5eea39f` on the console.

TAL.

**Comment by easyCZ** [comment]
> ested: When trying to sign-up, and clicking the "Continue" button on the questionaire, I get permission_denied] No access to Organization ID: 2a511bc8-cf3d-4911-a540-6169c5eea39f on the console.

It's the result of this needing a rebase on https://github.com/gitpod-io/gitpod/pull/18021

**Comment by geropl** [comment]
Re-tested, and saw the same error. At least the deploy job run on the latest commit ([b3f8296](https://github.com/gitpod-io/gitpod/pull/17993/commits/b3f82967173b0806d4127c02ce0556367403abee)), did not check the control plane though.

File: components/gitpod-db/src/typeorm/team-db-impl.ts (hunk 2 above)
Comment (svenefftinge) [reply components/gitpod-db/src/typeorm/team-db-impl.ts:191]: Can't we nest transactions if we use the same entity manager?

**Review by geropl — APPROVED** [review]
:+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

**Pull request merged**